packages/storage/amplify_storage_s3_dart/lib/src/amplify_storage_s3_dart_impl.dart

```diff
@@ -69,19 +111,46 @@ class AmplifyStorageS3Dart extends StoragePluginInterface {
       identityProvider: identityProvider,
     );
 
-    // TODO(HuiSF): create S3Client instance with
-    //  * credentialsProvider
-    //  * _prefixResolver
-    //  * other fields
-    // final credentialsProvider = authProviderRepo
-    //     .getAuthProvider(APIAuthorizationType.iam.authProviderToken);
+    final credentialsProvider = authProviderRepo
+        .getAuthProvider(APIAuthorizationType.iam.authProviderToken);
+
+    if (credentialsProvider == null) {
+      throw const StorageException(
+        'No credential provider found for Storage.',
+        recoverySuggestion:
+            'If you haven\'t already, please add amplify_auth_cognito plugin to your App.',
+      );
+    }
+
+    //`_dependencyManagerOverride` should be available only in unit tests
+    if (_dependencyManagerOverride == null) {
```

Comment:
Having logic that runs only in unit tests is a risky proposition and is one of the pitfalls of relying on unit tests to verify behavior. Injecting a dependency manager for tests is okay, but it should be used in exactly the same way it would be normally - otherwise the tests start to not actually verify the plugin behavior.

Calling `addInstance` multiple times will overwrite values from the previous call, so it is safe to call this in `configure` and then again in unit tests with the overridden dependency before calling plugin methods.